clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3523,6 +3523,23 @@ static bool InvalidWorkGroupSizeAttrs(const Expr *MGValue, const Expr *XDim,
            ZDimExpr->getResultAsAPSInt() != 1));
 }
 
+// If the [[intel::max_work_group_size(X, Y, Z)]] attribute is specified on
```

Comment:
> Does attribute spelling matter here? As in openCL vs SYCL. Doesn't `reqd_work_group_size` argument order depend on spelling?

No.

Reply:
Just for my understanding, can you please explain this?  When we add `reqd_work_group_size`, we check for openCL vs SYCL mode if declaration has an existing `num_simd_work_items` attribute i.e. the order matters. Why does the order in `reqd_work_group_size ` not matter when checking against `max_work_group_size`

Reply:
> Just for my understanding, can you please explain this? When we add `reqd_work_group_size`, we check for openCL vs SYCL mode if declaration has an existing `num_simd_work_items` attribute i.e. the order matters. Why does the order in `reqd_work_group_size ` not matter when checking against `max_work_group_size`

Here order does not matter because we are going to check all three arguments (x, y, z) of `reqd_work_group_size` with (x1, y1, z1) of `max_work_group_size` i.e x<=x1, y<=y1, z<=z1


For existing num_simd_work_items attribute i.e. the order matters. 
because We only consider work-group size of reqd_work_group_size attribute. 
Here work-group size is  last argument of  ``intel::reqd_work_group_size`` attribute
and first argument of `cl::reqd_work_group_size`` or ``__attribute__((reqd_work_group_size()))``
```
Introduce the num_simd_work_items attribute in conjunction with the reqd_work_group_size attribute. The num_simd_work_items attribute you specify must evenly divide the work-group size you specify for the reqd_work_group_size attribute.

we only consider work_size 
If the ``intel::reqd_work_group_size`` attribute is specified on a declaration
along with a ``intel::num_simd_work_items`` attribute, the work group size
(the last argument) must be evenly divisible by the argument specified
If the ``cl::reqd_work_group_size`` or ``__attribute__((reqd_work_group_size()))``
attribute is specified on a declaration along with a ``intel::num_simd_work_items``
attribute, the work group size (the first argument) must be evenly divisible by
the argument specified in the ``intel::num_simd_work_items`` attribute.
```

Reply:
@smanna12, so the spelling only matters for the reqd_work_group_size not for max_work_group_size?   Is that a good summary?

Reply:
Based on what @smanna12 said, my understanding is the order of `reqd_work_group_size ` does not matter when checking values against `max_work_group_size`, but it matters when checking against `num_simd_work_items`

Reply:
>>so the spelling only matters for the reqd_work_group_size not for max_work_group_size

Correct.

Reply:
Just to clarify - my concern didn't have to do with the `max_work_group_size`.  I understand `max_work_group_size` does not have different orderings. My concern was why why order of arguments in `reqd_work_group_size` mattered in some cases (for example - when comparing it to `num_simd_work_items`) and not in other cases (e.g. - when comparing it to `max_work_group_size`). Based on @smanna12's comment above, it looks like this is by design. I still don't fully understand why its designed this way but I trust @smanna12's judgement here!

Reply:
>>Based on @smanna12's comment above, it looks like this is by design.
 
Yes, The diagnostics are designed this way. My explanation is based on it. 
FYI
We do the same with max_global_work_dim and reqd_work_group_size attributes. Ordering does not matter

Reply:
> But when the reqd_work_group_size is prefixed with cl::, then the arguments are (z, y, x), aren't they?

Yes this was why I was confused. I wasn't sure if we needed to check third argument of `reqd_work_group_size `with first argument of  `max_work_group_size ` for `cl::` spellng

Reply:
> Yes this was why I was confused. I wasn't sure if we needed to check third argument of reqd_work_group_size with first argument of max_work_group_size for cl:: spelling.

FWIW, that's my worry as well. Triple-checking is always a good idea when dealing with this nosense.

> (FWIW, I too trust @smanna12's judgment, it is just that the specification seems to be designed to confuse me everytime.)

They are designed to be impenetrable to everyone and I look forward to the day when these attributes are deprecated so they can stop being such a time sink in the FE.

Reply:
> > Yes this was why I was confused. I wasn't sure if we needed to check third argument of reqd_work_group_size with first argument of max_work_group_size for cl:: spelling.
> 
> FWIW, that's my worry as well. Triple-checking is always a good idea when dealing with this nosense.

Existing test for this:
https://github.com/intel/llvm/blob/5b68889275ed765caed979413bf6b26da8d99113/clang/test/SemaSYCL/intel-max-work-group-size-device.cpp#L36-#L44

```
#ifdef TRIGGER_ERROR
struct DAFuncObj {
  [[intel::max_work_group_size(4, 4, 4)]] // expected-note {{conflicting attribute is here}}
  [[cl::reqd_work_group_size(8, 8, 4)]] // expected-error{{'reqd_work_group_size' attribute conflicts with 'max_work_group_size' attribute}} \
                                        // expected-warning{{attribute 'cl::reqd_work_group_size' is deprecated}} \
                                        // expected-note{{did you mean to use 'sycl::reqd_work_group_size' instead?}}
  void operator()() const {}
};
#endif // TRIGGER_ERROR
``

Reply:
>   [[intel::max_work_group_size(4, 4, 4)]] // expected-note {{conflicting attribute is here}}
>   [[cl::reqd_work_group_size(8, 8, 4)]] // expected-error{{'reqd_work_group_size' attribute conflicts with 'max_work_group_size' attribute}} \

This test isn't very helpful for this particular confusion since it is not apparent which '4' the '8' is being compared with.
Something like: max_work_group_size(1,2,3) with reqd_work_group_size(1,2,3) spelled with cl:: and intel:: prefixes might help.

Reply:
To explain my concern a bit better -  

If x = puppies, y = kittens and z = turtles

for `reqd_work_group_size `- my understanding is spellings are:

```
[[intel::reqd_work_group_size(NumberOfPuppies, NumberOfKittens, NumberOfTurtles)]] 
[[cl::reqd_work_group_size(NumberOfTurtles, NumberOfKittens, NumberOfPuppies)]]
```

For max_work_group_size there is only one spelling - 

`[[intel::max_work_group_size(MaximumAllowedPuppies, MaximumAllowedKittens, MaximumAllowedTurtles)]] `

If we aren't checking spelling for `reqd_work_group_size`, for `cl::`, aren't we checking `NumberOfTurtles `against `MaximumAllowedPuppies`?

Reply:
So this test is wrong right?

Reply:
> So this test is wrong right?

We give diagnostic for !(x<=x1)---> x>x1

Reply:
Right, but shouldn't 1 be compared to 3 here?

If `[[intel::reqd_work_group_size(X,Y,Z)]]` = `[[cl::reqd_work_group_size(Z,Y,X)]]` , I don't understand how 

```
[[intel::max_work_group_size(1,2,3)]]
[[cl::reqd_work_group_size(1,2,3)]] 
 void foo()() {}
```

is functionally equivalent to 

```
[[intel::max_work_group_size(1,2,3)]]
[[intel::reqd_work_group_size(1,2,3)]]
 void foo()() {}
```

As @AaronBallman suggested above, I think there is benefit to triple-checking required behavior with attribute authors. The existing behavior of this diagnostic might be incorrect.

Reply:
>>As @AaronBallman suggested above, I think there is benefit to triple-checking required behavior with attribute authors. The existing behavior of this diagnostic might be incorrect.

I already sent an email to double check about this to spec authors. i am waiting for his response.

Reply:
>>They are designed to be impenetrable to everyone and I look forward to the day when these attributes are deprecated so they can stop being such a time sink in the FE.

``__attribute__((reqd_work_group_size(1, 2, 3)))`` and ``[[cl::reqd_work_group_size(1, 2, 3)]]`` both have been deprecated in favor of 'sycl::reqd_work_group_size'

Reply:
Sorry my understanding was wrong. We need to check spellings and orderings of ``reqd_work_group_size`` attribute which takes default argument. I have forwarded offline email to all from @GarveyJoe